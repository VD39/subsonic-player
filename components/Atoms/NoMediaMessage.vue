<script setup lang="ts">
defineProps<{
  icon: string;
  message: string;
}>();
</script>

<template>
  <div :class="$style.noMediaMessage">
    <div :class="['column', 'centerItems', $style.content]">
      <component :is="icon" :size="50" />

      <p>{{ message }}</p>
    </div>
  </div>
</template>

<style module>
.noMediaMessage {
  position: relative;
  height: var(--width-height-100);
}

.content {
  position: absolute;
  inset: 50% auto auto 50%;
  transform: translate(-50%, -50%);
}
</style>
