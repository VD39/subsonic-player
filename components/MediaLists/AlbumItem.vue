<script setup lang="ts">
import ArtistsList from '@/components/TrackDetails/ArtistsList.vue';
import ImageLink from '@/components/Image/ImageLink.vue';

const props = defineProps<{
  album: Album;
}>();

const toLink = `/album/${props.album.id}`;
</script>

<template>
  <article>
    <ImageLink :to="toLink" :title="album.name" :image="album.image" />

    <p class="clamp clamp2">
      <NuxtLink :to="toLink" :title="`Go to album ${album.name}`">
        {{ album.name }}
      </NuxtLink>
    </p>

    <ArtistsList
      v-if="album.artists.length"
      :artists="album.artists"
      class="clamp clamp2"
    />
  </article>
</template>
