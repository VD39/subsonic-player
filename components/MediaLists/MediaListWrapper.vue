<script setup lang="ts">
withDefaults(
  defineProps<{
    rows?: string;
  }>(),
  {
    rows: '5',
  },
);
</script>

<template>
  <section :class="$style.mediaListWrapper" :style="{ '--loop-rows': rows }">
    <slot />
  </section>
</template>

<style module>
.mediaListWrapper {
  --loop-cols: calc(
    (var(--loop-max-width) - (var(--loop-columns) - 1) * var(--loop-grid-gap)) /
      var(--loop-columns)
  );

  position: relative;
  display: grid;
  grid-template-columns: repeat(auto-fill, minmax(var(--loop-cols), 1fr));
  gap: var(--loop-grid-gap);
  list-style: none;

  @media (width >= 420px) {
    --loop-columns: max(var(--loop-rows) - 3, 2);
  }

  @media (width >= 775px) {
    --loop-columns: max(var(--loop-rows) - 2, 2);
  }

  @media (width >= 1000px) {
    --loop-columns: max(var(--loop-rows) - 1, 2);
  }

  @media (width >= 1260px) {
    --loop-columns: var(--loop-rows);
  }
}
</style>
