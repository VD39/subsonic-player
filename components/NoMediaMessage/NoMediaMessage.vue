<script setup lang="ts">
defineProps<{
  icon: string;
  message: string;
}>();
</script>

<template>
  <div :class="$style.noMediaMessage">
    <div :class="$style.content">
      <component :is="icon" :size="50" />

      <p>{{ message }}</p>
    </div>
  </div>
</template>

<style module>
.noMediaMessage {
  position: relative;
  height: 100%;
}

.content {
  @mixin align-center;

  position: absolute;
  inset: 50% auto auto 50%;
  flex-direction: column;
  transform: translate(-50%, -50%);
}
</style>
