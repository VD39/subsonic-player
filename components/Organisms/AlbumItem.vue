<script setup lang="ts">
import ArtistsList from '@/components/Atoms/ArtistsList.vue';
import ButtonLink from '@/components/Atoms/ButtonLink.vue';
import SpinningLoader from '@/components/Atoms/SpinningLoader.vue';
import ImageLink from '@/components/Organisms/ImageLink.vue';

const props = defineProps<{
  album: Album;
  hideArtist?: boolean;
}>();

const { playTracks } = useAudioPlayer();
const { getAlbum } = useAlbum();

const loading = ref(false);

async function playAlbumTracks(albumId: string) {
  loading.value = true;

  const selectedAlbum = await getAlbum(albumId);

  if (selectedAlbum) {
    await playTracks(selectedAlbum.tracks);
  }

  loading.value = false;
}

const buttonProps = computed<ButtonProps>(() => ({
  icon: loading.value ? SpinningLoader : ICONS.play,
  text: `Play album ${props.album.name}`,
}));

const toLink = `/album/${props.album.id}`;
</script>

<template>
  <article>
    <div :class="$style.albumImageWrapper">
      <ImageLink
        :to="toLink"
        :title="`Go to album ${album.name}`"
        :image="album.image"
        class="mBS"
      />

      <ButtonLink
        :icon="buttonProps.icon"
        :class="['themeHoverButton', $style.buttonLink]"
        :title="buttonProps.text"
        @click="playAlbumTracks(album.id)"
      >
        {{ buttonProps.text }}
      </ButtonLink>
    </div>

    <p class="mBXS strong smallFont clamp2">
      <NuxtLink
        :to="toLink"
        class="link"
        :aria-label="`Go to album ${album.name}`"
      >
        {{ album.name }}
      </NuxtLink>
    </p>

    <ArtistsList
      v-if="!hideArtist && album.artists.length"
      :artists="album.artists"
      class="smallFont clamp2"
    />
  </article>
</template>

<style module>
.albumImageWrapper {
  position: relative;

  @media (hover: hover) {
    &:hover,
    &:focus {
      .buttonLink {
        --album-play-button-opacity: 1;
      }
    }
  }
}

.buttonLink {
  --album-play-button-opacity: 0;

  position: absolute;
  inset: auto var(--default-space) var(--default-space) auto;
  opacity: var(--album-play-button-opacity);
  transition: opacity var(--transition);
}
</style>
