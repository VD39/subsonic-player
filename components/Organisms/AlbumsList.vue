<script setup lang="ts">
import GridWrapper from '@/components/Atoms/GridWrapper.vue';
import NoMediaMessage from '@/components/Atoms/NoMediaMessage.vue';
import AlbumItem from '@/components/Organisms/AlbumItem.vue';

defineProps<{
  albums: Album[];
  hideArtist?: boolean;
}>();

defineEmits<{
  dragStart: [album: Album, event: DragEvent];
}>();
</script>

<template>
  <GridWrapper v-if="albums.length">
    <AlbumItem
      v-for="album in albums"
      :key="album.name"
      :album
      draggable="true"
      :hideArtist
      @dragstart="$emit('dragStart', album, $event)"
    />
  </GridWrapper>

  <NoMediaMessage
    v-else
    :icon="IMAGE_DEFAULT_BY_TYPE.album"
    message="No albums found."
  />
</template>
