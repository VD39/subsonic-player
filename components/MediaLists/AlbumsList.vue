<script setup lang="ts">
import NoMediaMessage from '@/components/NoMediaMessage/NoMediaMessage.vue';
import MediaListWrapper from './MediaListWrapper.vue';
import AlbumItem from './AlbumItem.vue';

defineProps<{
  albums: Album[];
  rows?: string;
}>();
</script>

<template>
  <MediaListWrapper v-if="albums.length" :row="rows">
    <AlbumItem v-for="album in albums" :key="album.name" :album="album" />
  </MediaListWrapper>
  <NoMediaMessage v-else icon="PhVinylRecord" message="No albums found." />
</template>
