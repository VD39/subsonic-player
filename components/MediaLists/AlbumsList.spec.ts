import type { VueWrapper } from '@vue/test-utils';
import { mount } from '@vue/test-utils';
import { getFormattedAlbumsMock } from '@/test/helpers';
import NoMediaMessage from '@/components/NoMediaMessage/NoMediaMessage.vue';
import MediaListWrapper from './MediaListWrapper.vue';
import AlbumItem from './AlbumItem.vue';
import AlbumsList from './AlbumsList.vue';

function factory(props = {}) {
  return mount(AlbumsList, {
    props: {
      albums: [],
      ...props,
    },
  });
}

describe('AlbumsList', () => {
  let wrapper: VueWrapper;

  beforeAll(() => {
    wrapper = factory();
  });

  it('matches the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  describe('when albums prop is an empty array', () => {
    it('does not show the MediaListWrapper component', () => {
      expect(wrapper.findComponent(MediaListWrapper).exists()).toBe(false);
    });

    it('shows the NoMediaMessage component', () => {
      expect(wrapper.findComponent(NoMediaMessage).exists()).toBe(true);
    });
  });

  describe('when albums prop is not an empty array', () => {
    beforeAll(() => {
      wrapper = factory({
        albums: getFormattedAlbumsMock(5),
      });
    });

    it('matches the snapshot', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('shows the MediaListWrapper component', () => {
      expect(wrapper.findComponent(MediaListWrapper).exists()).toBe(true);
    });

    it('shows the correct number of AlbumItem components', () => {
      expect(wrapper.findAllComponents(AlbumItem).length).toBe(5);
    });

    it('does not show the NoMediaMessage component', () => {
      expect(wrapper.findComponent(NoMediaMessage).exists()).toBe(false);
    });
  });
});
