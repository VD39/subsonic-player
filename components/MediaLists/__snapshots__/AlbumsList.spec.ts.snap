// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AlbumsList > matches the snapshot 1`] = `
"<div class="noMediaMessage">
  <div class="content">
    <phvinylrecord size="50"></phvinylrecord>
    <p>No albums found.</p>
  </div>
</div>"
`;

exports[`AlbumsList > when albums prop is not an empty array > matches the snapshot 1`] = `
"<section class="mediaListWrapper" style="--loop-rows: 5;">
  <article>
    <figure class="figure"><a title="Go to album-0">
        <div class="preloadImage"><span class="loader"><span class="visually-hidden">Loading image</span></span><img class="image" src="coverArt" alt="album-0" style="display: none;"></div>
      </a></figure>
    <p class="clamp clamp2"><a title="Go to album album-0">album-0</a></p>
    <ul class="commaList clamp clamp2">
      <li data-test-id="artist-list-item"><a>name</a></li>
      <li data-test-id="artist-list-item"><a>name1</a></li>
    </ul>
  </article>
  <article>
    <figure class="figure"><a title="Go to album-1">
        <div class="preloadImage"><span class="loader"><span class="visually-hidden">Loading image</span></span><img class="image" src="coverArt" alt="album-1" style="display: none;"></div>
      </a></figure>
    <p class="clamp clamp2"><a title="Go to album album-1">album-1</a></p>
    <ul class="commaList clamp clamp2">
      <li data-test-id="artist-list-item"><a>name</a></li>
      <li data-test-id="artist-list-item"><a>name1</a></li>
    </ul>
  </article>
  <article>
    <figure class="figure"><a title="Go to album-2">
        <div class="preloadImage"><span class="loader"><span class="visually-hidden">Loading image</span></span><img class="image" src="coverArt" alt="album-2" style="display: none;"></div>
      </a></figure>
    <p class="clamp clamp2"><a title="Go to album album-2">album-2</a></p>
    <ul class="commaList clamp clamp2">
      <li data-test-id="artist-list-item"><a>name</a></li>
      <li data-test-id="artist-list-item"><a>name1</a></li>
    </ul>
  </article>
  <article>
    <figure class="figure"><a title="Go to album-3">
        <div class="preloadImage"><span class="loader"><span class="visually-hidden">Loading image</span></span><img class="image" src="coverArt" alt="album-3" style="display: none;"></div>
      </a></figure>
    <p class="clamp clamp2"><a title="Go to album album-3">album-3</a></p>
    <ul class="commaList clamp clamp2">
      <li data-test-id="artist-list-item"><a>name</a></li>
      <li data-test-id="artist-list-item"><a>name1</a></li>
    </ul>
  </article>
  <article>
    <figure class="figure"><a title="Go to album-4">
        <div class="preloadImage"><span class="loader"><span class="visually-hidden">Loading image</span></span><img class="image" src="coverArt" alt="album-4" style="display: none;"></div>
      </a></figure>
    <p class="clamp clamp2"><a title="Go to album album-4">album-4</a></p>
    <ul class="commaList clamp clamp2">
      <li data-test-id="artist-list-item"><a>name</a></li>
      <li data-test-id="artist-list-item"><a>name1</a></li>
    </ul>
  </article>
</section>"
`;
